main.py

```diff
@@ -0,0 +1,112 @@
+from transformers import RobertaForMaskedLM
+from transformers.models.roberta.modeling_roberta import RobertaPooler
+from torchvision.models import resnet18
+import torch.nn as nn
+import torch 
+from transformers.data.data_collator import DataCollatorForLanguageModeling
+import pytorch_lightning as pl 
+from utils.model_utils import freeze_everything_except_top
+import random 
+
+class JeopardyModel(pl.LightningModule):
+    def __init__(self, config, pretrained_bert):
+        super().__init__()
+        self.model = RobertaForMaskedLM.from_pretrained('roberta-base')
+        freeze_everything_except_top(self.model, 3) # this seems to be too high for CCN 
+        self.word_embeddings = self.model.roberta.embeddings.word_embeddings
+        self.image_feature_extractor = resnet18(pretrained=False)
+        self.image_feature_extractor.fc = nn.Linear(512, self.word_embeddings.embedding_dim) 
+        self.alpha = config.model_params.alpha
+        self.op = config.optim_params
+        bs = config.optim_params.batch_size
+        self.idx = torch.tensor([bs-1] + [i for i in range(0, bs-1)])
+        self.sep_token = self.word_embeddings(torch.tensor([2])) # sep token is same as end token??
+        self.pooler = RobertaPooler(self.model.config) 
+
+    def forward_image(self, x):
+      return self.image_feature_extractor(x)
+
+    def forward(self, x):
+      out = self.model.roberta(**x)
+      # look into the input of pooler to see what it should be
+      cls_embed = self.pooler(out.last_hidden_state)
+      return cls_embed
+
+    def mlm_forward(self, x):
+      out = self.model(**x)
+      return out.loss
+
+    def training_step(self, batch, batch_idx):
+      loss = self.shared_step(batch)
+      self.log("train_loss", loss, prog_bar=True, sync_dist=True)
+      return loss 
+
+
+    def shared_step(self, batch):
+      '''
+        some notes:
+        - sep_token is the same as the eos_token in huggingface
+      '''
+      q, img1, img2, a = batch
+      
+      img1 = self.forward_image(img1)
+      img2 = self.forward_image(img2)
+      img1 = torch.unsqueeze(img1, dim=1)
+      img2 = torch.unsqueeze(img2, dim=1)
+      
+      img_attention_mask = torch.ones_like(a["attention_mask"]) # sep + img + eos token
+      attention_mask = torch.cat((q["attention_mask"], a["attention_mask"], img_attention_mask), dim=2)
+      
+      answer = self.word_embeddings(a["contrastive_input"])
+      sep_token = answer[:, :, -1] 
```

Comment:
why is this the sep_token?

Reply:
could we talk about this when we meet? I think I'm being inefficient on memory usage here